### DISCUSSION TITLE
Tracking humans over long durations (months)

### DISCUSSION THREAD

**Discussion by louis030195** [ultralytics/ultralytics#5171] [Q&A]
Hey do you have any recommendation do track human over long durations, like months?

I'm not super familiar with object tracking, but I assume the tracker compute a kind of embeddings of each object and keep a dictionary in memory, upon detecting new object check if the embedding is close enough to existing object's embedding in memory it will use its ID?

So to do this over long durations I would need to save this dict on the filesystem probably in a persistent storage and use it every time I run the tracking.

Any thoughts? (any code would be super helpful, otherwise will dig by myself) 🙏

**Comment by louis030195** [comment]
Also I see that it actually fails to track "uniquely" humans like a kind of face recognition but for the whole body, I guess tracking uses a simple heuristic? Do I have to train a custom face rec for the body to achieve that?

**Reply by glenn-jocher** [reply]
@louis030195 hello! Tracking humans over long durations, such as months, presents unique challenges due to changes in appearance, lighting, and scene context. YOLOv8's tracking mode is designed for real-time applications and may not be directly suitable for long-term tracking without additional mechanisms.

For long-term tracking, you would indeed need a way to persistently store and update object representations. However, YOLOv8's tracking is not inherently designed for long-term identity preservation over such extended periods, especially when the subjects may not be continuously in the camera's view.

Regarding your observation about the tracking mechanism, YOLOv8's tracker typically uses spatial and appearance cues for short-term tracking. It does not employ complex embeddings that could handle significant appearance changes over time, which would be more akin to a re-identification system.

For long-term tracking with unique identification, you might consider integrating a separate re-identification model that can generate robust embeddings for each individual. This would involve training a model on a dataset with labeled individuals across various conditions to learn distinctive features.

While we don't provide code for such a complex system, you can explore the Ultralytics Docs for more information on the capabilities of YOLOv8 and how it might be extended or integrated with other systems for your specific use case. Good luck with your project! 🚀

**Comment by louis030195** [comment]
bump

**Reply by glenn-jocher** [reply]
@louis030195 hello! Tracking humans over long durations, such as months, is indeed a challenging task due to changes in appearance, lighting, and scene context. While YOLOv8 provides real-time object tracking capabilities, it's primarily designed for short-term tracking where objects are continuously visible and the environment does not change drastically.

For long-term tracking, you would need a system that can handle re-identification (Re-ID) across different sessions and possibly different cameras. This typically involves generating robust embeddings for each individual that can be matched over time, even with changes in clothing, lighting, or camera perspective.

The tracking mode in YOLOv8 can be a starting point for real-time tracking, but for long-term tracking, you would need to integrate a Re-ID system that can store and compare embeddings over time. This would involve saving embeddings to persistent storage, as you mentioned, and implementing a mechanism to match new detections with stored embeddings.

For more information on the tracking capabilities of YOLOv8, you can refer to the Track mode documentation. However, please note that for your specific use case, additional research and development beyond the scope of YOLOv8's built-in features will be necessary. Good luck with your project! 🍀

**Comment by PiyushKamboj1503** [comment]
Hello @louis030195 , the trackers like Botsort and ByteTrack won't go a long way without configuration of them. You can do so by changing the parameters available in the botsort.yaml and bytetrack.yaml. The point here to note is that when you are tracking human movements over months, it is feasible to generate the output in .csv. The Yolov8 framework will give you txt files of the detections which you can continously convert. Also, with a little modification in the code,  you can set the name of the detections as date and time.txt which allows you to have a better csv. Also, to reinforce your model better, you can save_crop all your detections and can pass them through opencv harcascade classifiers and lbph recognizers to create a lightweight recognition package that continously trains on your saved_crops and once you get a hit(recognition), you can mention the same in your csv by editing. Hope this helps.
